codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -319,13 +327,86 @@ private int allocateBytesForTree(Http2Stream parent, int connectionWindow) {
             }
             connectionWindow = nextConnectionWindow;
             totalWeight = nextTotalWeight;
-            tail = nextTail;
         }
 
         return bytesAllocated;
     }
 
     /**
+     * Utility class used by the priority algorithm to manage multiple iterations over the children of a given stream.
+     * This class will try to avoid copying the {@link Collection} returned by the parent {@link
+     * Http2Stream#children()}} by first iterating over the collection directly. As streams are found that require
+     * additional processing, they will be added to the next iteration buffer via {@link
+     * #addToNextIteration(Http2Stream)}. The next call to {@link #iterator()} will return an {@link Iterator} that will
+     * iterate only over those streams which were added back to the cache.
+     */
+    private final class ChildCache implements Iterable<Http2Stream> {
+        final Collection<? extends Http2Stream> children;
+        int nextTail;
+        Http2Stream[] childArray = null;
+
+        /**
+         * Creates the cache and initializes it to iterate over all children of the given parent in the next
+         * iteration.
+         */
+        ChildCache(Http2Stream parent) {
+            this.children = parent.children();
+            nextTail = children.size();
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public Iterator<Http2Stream> iterator() {
+            try {
+                if (nextTail == 0) {
+                    // Nothing to iterate over.
+                    return Collections.emptyIterator();
+                }
+
+                if (childArray != null) {
+                    // We've allocated the child array, iterate over it up to the specified tail position.
+                    return Iterators.newIterator(childArray, 0, nextTail);
+                }
+
+                if (nextTail != children.size()) {
```

Comment:
This seems more like a programming error...should we make this an `assert` with a message perhaps?